### PR TITLE
Add xparl status & @remote_class(max_memory)

### PR DESCRIPTION
**Pull request by fuyw**

### CODE REVIEW

File: parl/remote/master.py [hunk 1]
```diff
@@ -37,10 +37,11 @@ class Master(object):
 
     .. code-block:: python
 
-        xparl start --port localhost:1234
+        xparl start --port localhost:1234 --monitor_port 5005
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/master.py]: We should not mention monitor_port here. The usage here should be as compact as possible. Please add monitor_port at our documentation.

File: parl/remote/master.py [hunk 2]
```diff
@@ -37,10 +37,11 @@ class Master(object):
...
 
     At the same time, a local worker will be started and connect to the
-    master node.
+    master node. And a cluster monitor is start at `localhost:5005` to
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/master.py]: Remove detail about the monitor here.

File: parl/remote/scripts.py [hunk 3]
```diff
@@ -86,10 +86,18 @@ def cli():
     type=int,
     help="Set number of cpu manually. If not set, it will use all "
     "cpus of this machine.")
-def start_master(port, cpu_num):
+@click.option(
+    "--monitor_port", help="The port to start cluster monitor", type=str)
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/scripts.py]: The port to start a cluster monitor

File: docs/parallel_training/features.rst [hunk 4]
```diff
@@ -0,0 +1,53 @@
+Features
```
Comment (TomorrowIsAnOtherDay) [inline docs/parallel_training/features.rst]: rename this file to overview.rst.

File: docs/parallel_training/features.rst [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+ including cpu usage, memory usage, average cpu loads and file executable
+ paths. When any error happens, user can soon locate the bugs.
+
+Board compatibility
+^^^^^^^^^^^^^^^^^^^
+| PARL is compatible with any other frameworks, like tensorflow, pytorch or
+ mxnet. By adding `@parl.remote_class` decorator to their code, users can
+ easily convert their code to distributed computation.
+
+Advantages
```
Comment (TomorrowIsAnOtherDay) [inline docs/parallel_training/features.rst]: Advantages----> Why PARL

File: docs/parallel_training/features.rst [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+| `@parl.remote_class` enable us to achieve real multi-thread computation
+ efficiency without modifying our codes. As shown in figure (a), python's
+ original multi-thread computation performs poorly due to the limitation
+ of the GIL, while PARL empowers us to realize real parallel computation
+ efficiency.
+
+Convenient visualization
+^^^^^^^^^^^^^^^^^^^^^^^^
+| PARL also provides a web monitor to watch the cluster status. We can use
+ `xparl status` command to visualize the details of our cluster status,
```
Comment (TomorrowIsAnOtherDay) [inline docs/parallel_training/features.rst]: `xparl status` -> \`\`xparl status``

File: docs/parallel_training/features.rst [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+High performance
+^^^^^^^^^^^^^^^^
+| `@parl.remote_class` enable us to achieve real multi-thread computation
+ efficiency without modifying our codes. As shown in figure (a), python's
+ original multi-thread computation performs poorly due to the limitation
+ of the GIL, while PARL empowers us to realize real parallel computation
+ efficiency.
+
+Convenient visualization
```
Comment (TomorrowIsAnOtherDay) [inline docs/parallel_training/features.rst]: Computation Resources Monitor.
PARL provides a web monitor to watch the status of any resources connected to the cluster.

File: docs/parallel_training/features.rst [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^
+| PARL also provides a web monitor to watch the cluster status. We can use
+ `xparl status` command to visualize the details of our cluster status,
+ including cpu usage, memory usage, average cpu loads and file executable
+ paths. When any error happens, user can soon locate the bugs.
```
Comment (TomorrowIsAnOtherDay) [inline docs/parallel_training/features.rst]: file executable paths  -> the path of each task.

File: parl/remote/remote_decorator.py [hunk 9]
```diff
@@ -55,6 +55,15 @@ def step(self):
...
         actor.step()
 
+        # Set maximum memory usage for each object 
+        @remote_class(max_memory=300)
+        class LimitedActor(object):
+           ...
+
+    Args:
+        max_memory (float): Max memory each job process can use, the unit is in
+                            MB and default value is none.
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/remote_decorator.py]: default value is none(unlimited).

File: parl/remote/remote_decorator.py [hunk 10]
```diff
@@ -55,6 +55,15 @@ def step(self):
         actor = Actor()
         actor.step()
 
+        # Set maximum memory usage for each object 
+        @remote_class(max_memory=300)
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/remote_decorator.py]: parl.remote_class(max_memroy=300)

File: parl/remote/remote_decorator.py [hunk 11]
```diff
@@ -55,6 +55,15 @@ def step(self):
         actor = Actor()
         actor.step()
 
+        # Set maximum memory usage for each object
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/remote_decorator.py]: Set maximum memory usage to 300 MB for each object.

File: parl/remote/job.py [hunk 12]
```diff
@@ -45,9 +47,15 @@ def __init__(self, worker_address):
         """
         Args:
             worker_address(str): worker_address for sending job information(e.g, pid)
+
+        Attributes:
+            pid (int): Job process ID.
+            max_memory (float): Max memory (MB) can be used by this job process.
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/job.py]: Maximum memory (MB) can be used by each remote instance.

File: parl/remote/scripts.py [hunk 13]
```diff
@@ -168,9 +183,40 @@ def stop():
...
+            master_ip = master_address.split(':')[0]
+            monitor_address = "{}:{}".format(master_ip, monitor_port)
+            socket = ctx.socket(zmq.REQ)
+            socket.connect('tcp://{}'.format(master_address))
+            socket.send_multipart([STATUS_TAG])
+            cluster_info = to_str(socket.recv_multipart()[1])
+            msg = """
+            # Cluster {} {}
+
+            # If you want to monitor cluster status, visit: http://{}
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/scripts.py]: monitor -> check

File: docs/parallel_training/overview.rst [hunk 14]
```diff
@@ -0,0 +1,53 @@
+Features
+========
+
+Highlights
+##########
```
Comment (TomorrowIsAnOtherDay) [inline docs/parallel_training/overview.rst]: Remove this subtitle

File: docs/parallel_training/overview.rst [hunk 15]
```diff
@@ -0,0 +1,53 @@
+Features
+========
```
Comment (TomorrowIsAnOtherDay) [inline docs/parallel_training/overview.rst]: Replace Features with Summary.

File: docs/parallel_training/overview.rst [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+
+High performance
+^^^^^^^^^^^^^^^^
+| `@parl.remote_class` enable us to achieve real multi-thread computation
+ efficiency without modifying our codes. As shown in figure (a), python's
+ original multi-thread computation performs poorly due to the limitation
+ of the GIL, while PARL empowers us to realize real parallel computation
+ efficiency.
+
+Computation resources monitor
```
Comment (TomorrowIsAnOtherDay) [inline docs/parallel_training/overview.rst]: Web UI for computation resources

File: docs/parallel_training/overview.rst [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+| PARL provides a web monitor to watch the status of any resources connected
```
Comment (TomorrowIsAnOtherDay) [inline docs/parallel_training/overview.rst:21]: PARL provides a web monitor to watch the status of any resources connected
 to the cluster. We can use ``xparl status`` command to visualize the details
 of our cluster status, including cpu usage, memory usage, average cpu loads
 and the path of each task. When any error happens, user can soon locate the bugs.

-----

 PARL provides a web monitor to watch the status of any resources connected
 to the cluster. Users can view the cluster status at a WEB UI. It shows the detailed information for each worker(e.g, memory used) and each task submitted.

File: docs/parallel_training/overview.rst [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+ to the cluster. We can use ``xparl status`` command to visualize the details
+ of our cluster status, including cpu usage, memory usage, average cpu loads
+ and the path of each task. When any error happens, user can soon locate the bugs.
+
+Board compatibility
+^^^^^^^^^^^^^^^^^^^
+| PARL is compatible with any other frameworks, like tensorflow, pytorch or
```
Comment (TomorrowIsAnOtherDay) [inline docs/parallel_training/overview.rst]: Our framework for distributed training is compatible with ...

File: parl/remote/cluster_monitor.py [hunk 19]
```diff
@@ -106,6 +106,21 @@ def drop_cluster_status(self, client_address):
         self.status['clients'].pop(client_address)
         self.lock.release()
 
+    def get_status_info(self):
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/cluster_monitor.py:109]: Can you add a test script for this part of code?

File: parl/remote/job.py [hunk 20]
```diff
@@ -119,6 +127,16 @@ def _create_sockets(self):
             zmq.RCVTIMEO, remote_constants.HEARTBEAT_TIMEOUT_S * 1000)
         self.kill_job_socket.connect("tcp://{}".format(kill_job_address))
 
+    def _check_used_memory(self):
+        """Return current memory usage of the job process."""
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/job.py]: Check if the memory used by this job exceeds `self.max_memory`

File: parl/remote/job.py [hunk 21]
```diff
@@ -119,6 +127,16 @@ def _create_sockets(self):
...
+        stop_job = 'False'
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/job.py]: Why don't you use False(bool) here ?

File: parl/remote/master.py [hunk 22]
```diff
@@ -178,8 +179,13 @@ def _receive_message(self):
             self.client_socket.send_multipart(
                 [remote_constants.NORMAL_TAG, status])
 
-        elif tag == remote_constants.WORKER_INITIALIZED_TAG:
+        elif tag == remote_constants.STATUS_TAG:
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/master.py:183]: Please add a script testing this part of code.

File: parl/remote/remote_decorator.py [hunk 23]
```diff
@@ -63,137 +72,149 @@ def step(self):
...
-                return ret
+                return wrapper
 
-            return wrapper
+        return RemoteWrapper
 
-    return RemoteWrapper
+    max_memory = kwargs.get('max_memory')
+    if len(args) == 1 and callable(args[0]):
+        max_memory = None
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/remote_decorator.py]: Please remove this line.

File: parl/remote/scripts.py (hunk 13 above)
Comment (TomorrowIsAnOtherDay) [reply parl/remote/scripts.py]: visit -> view

**Review by TomorrowIsAnOtherDay — CHANGES_REQUESTED** [review]

File: docs/parallel_training/overview.rst [hunk 24]
```diff
@@ -0,0 +1,53 @@
+Features
+========
+
+Highlights
+##########
+
+Simple usage
```
Comment (TomorrowIsAnOtherDay) [inline docs/parallel_training/overview.rst]: Easy-to-use

File: docs/parallel_training/overview.rst [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+
+Why PARL
+########
+
+High throughput
+^^^^^^^^^^^^^^^
+| PARL uses a point-to-point connection for network communication in the
+ cluster. Unlike other framework like RLlib which replies on redis for
+ communication, PARL is able to achieve much higher throughput. The results
+ can be found in figure (b).
```
Comment (TomorrowIsAnOtherDay) [inline docs/parallel_training/overview.rst]: With the same implementation in IMPALA, PARL achieved an increase of 160% on data throughout over Ray(RLlib).

File: docs/parallel_training/overview.rst [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+
+Automatic deployment
+^^^^^^^^^^^^^^^^^^^^
+| Unlike other parallel framework like Ray, which fails to import modules from
```
Comment (TomorrowIsAnOtherDay) [inline docs/parallel_training/overview.rst]: pls remove `like Ray`.  It seems a bit aggressive.

File: docs/parallel_training/overview.rst [hunk 27]
```diff
@@ -0,0 +1,53 @@
+Features
+========
+
+Highlights
+##########
+
+Simple usage
+^^^^^^^^^^^^
+| With a single `@parl.remote_class` decorator, we can access cluster's remote
```
Comment (TomorrowIsAnOtherDay) [inline docs/parallel_training/overview.rst]: With a single `@parl.remote_class` decorator, users can implement parallel training easily, and  not having to care about stuff of multi-processes, network communication.

File: parl/remote/cluster_monitor.py (hunk 19 above)
Comment (fuyw) [reply parl/remote/cluster_monitor.py:109]: We already have one in tests/remote_monitor_test_x.py

File: parl/remote/job.py (hunk 21 above)
Comment (fuyw) [reply parl/remote/job.py]: This `stop_job` flag will be sent to client, and zmq socket only receive a byte like this `to_byte(str(Fasle))`

Comment (TomorrowIsAnOtherDay) [reply parl/remote/job.py]: Yes, I know that only byte can be sent through ZMQ, but it's curious to return a string representing the test result. You may return a bool here and transfer it into a string before sending it out.

File: docs/parallel_training/overview.rst [hunk 28]
```diff
@@ -24,7 +23,8 @@ Computation resources monitor
 | PARL provides a web monitor to watch the status of any resources connected
  to the cluster. We can use ``xparl status`` command to visualize the details
```
Comment (TomorrowIsAnOtherDay) [inline docs/parallel_training/overview.rst]: We can **also** use

File: docs/parallel_training/overview.rst [hunk 29]
```diff
@@ -40,11 +40,12 @@ High throughput
...
  communication, PARL is able to achieve much higher throughput. The results
- can be found in figure (b).
+ can be found in figure (b). With the same implementation in IMPALA, PARL
+ achieved an increase of 160% on data throughout over Ray(RLlib).
 
 Automatic deployment
 ^^^^^^^^^^^^^^^^^^^^
-| Unlike other parallel framework like Ray, which fails to import modules from
+| Unlike other parallel frameworks which fail to import modules from
  external files. PARL will automatically package all related files and send
```
Comment (TomorrowIsAnOtherDay) [inline docs/parallel_training/overview.rst]: files. -> file,

File: parl/remote/client.py [hunk 30]
```diff
@@ -195,8 +195,14 @@ def _create_job_monitor(self, job_heartbeat_socket):
             try:
                 job_heartbeat_socket.send_multipart(
                     [remote_constants.HEARTBEAT_TAG])
-                _ = job_heartbeat_socket.recv_multipart()
-                time.sleep(remote_constants.HEARTBEAT_INTERVAL_S)
+                job_message = job_heartbeat_socket.recv_multipart()
+                if to_str(job_message[1]) == 'True':
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/client.py]: It's hard to maintain the code written in this way. You should define a variable representing the meaning of `job_message[1]` and then check whether it is `True`.

**Review by TomorrowIsAnOtherDay — APPROVED** [review]

**Pull request merged**